nvflare/app_opt/job_launcher/docker_launcher.py

```diff
@@ -0,0 +1,195 @@
+            if container:
+                self.logger.debug(f"container state: {container.status}, job states to enter: {job_states_to_enter}")
+                if container.status in job_states_to_enter:
+                    return True
+                elif timeout is not None and time.time() - starting_time > timeout:
+                    return False
+                time.sleep(1)
+            else:
+                return False
+
+
+class DockerJobLauncher(JobLauncherSpec):
+    def __init__(self, workspace: str, mount_path: str, network: str, timeout=None):
+        super().__init__()
+
+        self.workspace = workspace
+        self.mount_path = mount_path
+        self.network = network
+        self.timeout = timeout
+
+    def launch_job(self, job_meta: dict, fl_ctx: FLContext) -> JobHandleSpec:
+        self.logger.debug("DockerJobLauncher start to launch job")
+        job_image = extract_job_image(job_meta, fl_ctx.get_identity_name())
+
+        workspace_obj: Workspace = fl_ctx.get_prop(FLContextKey.WORKSPACE_OBJECT)
+        job_id = job_meta.get(JobConstants.JOB_ID)
+        app_custom_folder = workspace_obj.get_app_custom_dir(job_id)
+
+        python_path = f"{app_custom_folder}:$PYTHONPATH" if app_custom_folder != "" else "$PYTHONPATH"
+        job_name, cmd = self.get_command(job_meta, fl_ctx)
+        command = f' /bin/bash -c "export PYTHONPATH={python_path};{cmd}"'
+        self.logger.info(f"Launch image:{job_image}, run command: {command}")
+
+        client = docker.from_env()
+        try:
+            container = client.containers.run(
+                job_image,
+                command=command,
+                name=job_name,
+                network=self.network,
+                detach=True,
+                # remove=True,
+                volumes={
+                    self.workspace: {
+                        "bind": self.mount_path,
+                        "mode": "rw",
+                    },
+                },
+                # ports=ports,  # Map container ports to host ports (optional)
+            )
+            self.logger.info(f"Launch the job in DockerJobLauncher using image: {job_image}")
+
+            handle = DockerJobHandle(container)
+            try:
+                if handle.enter_states([DOCKER_STATE.RUNNING], timeout=self.timeout):
+                    return handle
+                else:
+                    handle.terminate()
+                    return None
+            except:
+                handle.terminate()
+                return None
+
+        except docker.errors.ImageNotFound:
+            self.logger.error(f"Failed to launcher job: {job_id} in DockerJobLauncher. Image '{job_image}' not found.")
+            return None
+        except docker.errors.APIError as e:
+            self.logger.error(f"Error starting container: {e}")
+            return None
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.GET_JOB_LAUNCHER:
```

Comment:
fed_utils get_job_launcher() fire the GET_JOB_LAUNCHER to get the job launcher to use.

Reply:
server_engine.py and client_executor.py will call the get_job_launcher() and fire the GET_JOB_LAUNCHER event to get the proper job launcher to launch the SJ and CJ before launching the job run.

Reply:
The reason to use event mechanism is for decoupling of the "core" from the launchers. The core (client_executor) only fires the event, and it's up to others listen to the event and offer them if they can launch the job.

Event mechanism is our primary approach fro decopling.

Reply:
The timing is that when the "core" (server_engine or client_executor) is about to launch the job, it will fire the event. It does so by calling fed_utils get_job_luncher() function, which actually fires the event.

Are you questioning the event name GET_JOB_LAUNCHER?

Reply:
^ @yhwen this comment is the only thing left

Reply:
renamed the event to BEFORE_JOB_LAUNCH.